Source/Diagnostics/ReducedDiags/FieldProbe.cpp

```diff
@@ -0,0 +1,293 @@
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(interp_order <= WarpX::nox ,
+                                     "Field probe interp_order should be lower than algo.particle_shape");
+    // resize data array
+    m_data.resize(noutputs*nLevel, 0.0_rt);
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs{m_path + m_rd_name + "." + m_extension, std::ofstream::out};
+            // write header row
+            int c = 0;
+            ofs << "#";
+            ofs << "[" << c++ << "]step()";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]time(s)";
+            for (int lev = 0; lev < nLevel; ++lev)
+            {
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_Ex_lev" + std::to_string(lev) + " (V/m)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_Ey_lev" + std::to_string(lev) + " (V/m)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_Ez_lev" + std::to_string(lev) + " (V/m)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_|E|_lev" + std::to_string(lev) + " (V/m)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_Bx_lev" + std::to_string(lev) + " (T)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_By_lev" + std::to_string(lev) + " (T)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_Bz_lev" + std::to_string(lev) + " (T)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]probe_|B|_lev" + std::to_string(lev) + " (T)";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+}
+// end constructor
+
+// function that computes field values at probe position
+void FieldProbe::ComputeDiags (int step)
+{
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get a reference to WarpX instance
+    auto & warpx = WarpX::GetInstance();
+
+    // get number of level
+    const auto nLevel = warpx.finestLevel() + 1;
+
+    // loop over refinement levels
+    for (int lev = 0; lev < nLevel; ++lev)
+    {
+
+        const amrex::Geometry& gm = warpx.Geom(lev);
+        const auto prob_lo = gm.ProbLo();
+        const auto prob_hi = gm.ProbHi();
+
+#if (AMREX_SPACEDIM == 2)
+        const bool probe_in_domain = x_probe >= prob_lo[0] and x_probe <= prob_hi[0] and
+                                     z_probe >= prob_lo[1] and z_probe <= prob_hi[1];
+#else
+        const bool probe_in_domain = x_probe >= prob_lo[0] and x_probe <= prob_hi[0] and
+                                     y_probe >= prob_lo[1] and y_probe <= prob_hi[1] and
+                                     z_probe >= prob_lo[2] and z_probe <= prob_hi[2];
+#endif
+        amrex::Real fp_Ex, fp_Ey, fp_Ez, fp_E, fp_Bx, fp_By, fp_Bz, fp_B;
```

Comment:
If you decide to have both the checks for `probe_in_domain` and `probe_proc != -1` (see below), you could initialize these values to 0._rt, and then you don't have to do anything more in case one of these checks fail.

Reply:
Now I'm actually wondering if I should have an assert for both these two cases so that the code crashes if the user specifies an invalid position fr the probe. What do you think?

Reply:
Hmm, I'm not sure. Maybe I would leave it as it is (return 0s) because if we use this probe with a moving window, it's possible that the probe is inside the domain for some timesteps and outside for others.

Reply:
Fair point!